content/blog/upkeep-testthat-3/index.Rmd

```diff
@@ -0,0 +1,142 @@
+
+If you haven't heard of testthat 3e yet, the [testthat article introducing the 3rd edition](https://testthat.r-lib.org/articles/third-edition.html) is a great place to start. It outlines all the changes this edition brings.
+
+While you can continue to use testthat's previous behaviour, it's a good idea to upgrade so that you can make use of handy new features. As some of the changes may break your tests, you might have been putting that off though? You would not be alone in that! Several tidymodels packages still have to make the jump but I recently upgraded [dials](https://github.com/tidymodels/dials/) and [censored](https://github.com/tidymodels/censored/) to testthat edition 3. Here is what I did and learned along the way.
+
+### Workflow to upgrade
+
+The testthat article tells you how you can opt in to the new edition, and about major changes: deprecations, how messages and warnings are handled, and how comparisons of objects are made.
+
+This already gives you the main guidance for a workflow:
+
+1. Activate edition 3.
+2. Remove or replace deprecated functions.
+3. If your output got noisy, quiet things down as needed.
+4. The harder part: think about what it means if things are not "all equal" anymore.
+
+### Activation `r emo::ji("rocket")`
+
+To activate you need to do two things in the DESCRIPTION - or you can let `usethis::use_testthat(3)` do it for you:
+
+- Increase the testthat version to `>= 3.0.0`.
+- Set the `Config/testthat/edition` field to `3`.
+
+### Moving on from deprecations `r emo::ji("sparkles")`
+
+The article on testthat 3e contains a [list of deprecated functions](https://testthat.r-lib.org/articles/third-edition.html#deprecations) together with their successors. You can work your way through it, searching for the deprecated function and then replacing it with the most suitable alternative. The first one that list is `context()` as testthat will use the file name instead, ensuring that context and file name are in sync. As such, `context()` does not have a replacement and my first commit after activating the third edition was to remove all calls to `context()`, followed by replacing other deprecated functions and arguments.
+
+![A list of commits starting with "require testthat 3e, followed by removing `context()` and other deprecated functions"](commits.png)
+
+### Warnings and messages `r emo::ji("shushing_face")`
+
+Edition 3 handles warnings and messages differently than edition 2: `expect_warning()` captures at most one warning so if your code generates multiple warnings, they will bubble up now. Messages were previously silently ignored, now they also bubble up. That means the output may be a lot noisier after switching to edition 3. If the warnings or messages are important, you should explicitly capture them. Otherwise you can suppress them to clean up the output and make it easier to focus on what's important. Again, the testthat article has good examples for how to do either.
```

Comment:
```suggestion
testthat edition 3 handles warnings and messages differently than edition 2: `expect_warning()` captures at most one warning, so if your code generates multiple warnings, they will bubble up now. Messages were previously silently ignored, now they also bubble up. That means the output may be a lot noisier after switching to edition 3. If the warnings or messages are important, you should explicitly capture them. Otherwise you can suppress them to clean up the output and make it easier to focus on what's important. Again, the testthat article has good examples for how to do either.
```